Add symbol guard counts instrumentation

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119349
* #119347
* __->__ #119290

This helps us understand if there are symbols which are extremely hot
(i.e., have a lot of guards mentioning them).  Extremely hot symbols are
candidates for being turned static.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang committed:
Add symbol guard counts instrumentation

This helps us understand if there are symbols which are extremely hot
(i.e., have a lot of guards mentioning them).  Extremely hot symbols are
candidates for being turned static.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3086,6 +3093,9 @@ def issue_guard(guard: ShapeGuard) -> None:
                 "free_symbols": sum(1 for v in symbol_to_source.values() if v),
+                # The keys are meaningless from an aggregate perspective, so
+                # don't include them.  Biggest first.
+                "symbol_guard_counts": sorted(self.symbol_guard_counter.values(), reverse=True),
```
Won't this not tell us much, unless we also include the dynamo source that the "culprit" symbols map to?

Or are you just trying to use this for now to get a rough sense of the distribution

ezyang replied:
It doesn't tell us much at all, but that's OK, I want to look at this data in aggregate over all jobs so even knowing the source in this context wouldn't be very meaningful.

ezyang committed:
Update on "Add symbol guard counts instrumentation"


This helps us understand if there are symbols which are extremely hot
(i.e., have a lot of guards mentioning them).  Extremely hot symbols are
candidates for being turned static.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Add symbol guard counts instrumentation"


This helps us understand if there are symbols which are extremely hot
(i.e., have a lot of guards mentioning them).  Extremely hot symbols are
candidates for being turned static.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge
